### ISSUE TITLE
[Binary releases] Create a script that will containerise the DEB build process

### ISSUE THREAD

**Issue by paulkatsoulakis** [netdata/netdata#5968]
##### Feature idea summary
Document the required activities that will generate a DEB package for a distribution, then create a scriptlet that will establish a container and execute the workflow within the container environment.

The scriptlet will determine the architecture and distro details, that will then determine the distro the DEB package is for.

References #5963

**Comment by paulkatsoulakis** [comment]
We have finalised our design for this, we will be utilising package cloud as our infrastructure provider for the packages. Travis CI will be preparing the RPM and DEB files and will deploy them there for our users to parse.

The repository is: https://packagecloud.io/netdata/netdata and instructions on how to set it up on your system can be found here: https://packagecloud.io/netdata/netdata/install

Here's a sample repository setup within a CentOS 7 OS (I have masked security related info)

```
[root@XXXXXX netdata]# curl -s https://packagecloud.io/install/repositories/netdata/netdata/script.rpm.sh | sudo bash
bash: sudo: command not found
[root@XXXXXX netdata]# curl -s https://packagecloud.io/install/repositories/netdata/netdata/script.rpm.sh | bash     
Detected operating system as centos/7.
Checking for curl...
Detected curl...
Downloading repository file: https://packagecloud.io/install/repositories/netdata/netdata/config_file.repo?os=centos&dist=7&source=script
done.
Installing pygpgme to verify GPG signatures...
Loaded plugins: fastestmirror, ovl
Determining fastest mirrors
 * base: ftp.otenet.gr
 * extras: ftp.otenet.gr
 * updates: ftp.otenet.gr
base                                                                                                                                                             | 3.6 kB  00:00:00     
extras                                                                                                                                                           | 3.4 kB  00:00:00     
netdata_netdata-source/signature                                                                                                                                 |  819 B  00:00:00     
Retrieving key from https://packagecloud.io/netdata/netdata/gpgkey
Importing GPG key XXXXXXXX:
 Userid     : "https://packagecloud.io/firehol/netdata (https://packagecloud.io/docs#gpg_signing) <support@packagecloud.io>"
 Fingerprint: XXXXXXXXXXXXX
 From       : https://packagecloud.io/netdata/netdata/gpgkey
netdata_netdata-source/signature                                                                                                                                 |  951 B  00:00:03 !!! 
updates                                                                                                                                                          | 3.4 kB  00:00:00     
(1/2): extras/7/x86_64/primary_db                                                                                                                                | 201 kB  00:00:00     
(2/2): updates/7/x86_64/primary_db                                                                                                                               | 4.2 MB  00:00:04     
netdata_netdata-source/primary                                                                                                                                   |  175 B  00:00:01     
Package pygpgme-0.3-9.el7.x86_64 already installed and latest version
Nothing to do
Installing yum-utils...
Loaded plugins: fastestmirror, ovl
Loading mirror speeds from cached hostfile
 * base: ftp.otenet.gr
 * extras: ftp.otenet.gr
 * updates: ftp.otenet.gr
Package yum-utils-1.1.31-50.el7.noarch already installed and latest version
Nothing to do
Generating yum cache for netdata_netdata...
Importing GPG key XXXXXXXXXX:
 Userid     : "https://packagecloud.io/firehol/netdata (https://packagecloud.io/docs#gpg_signing) <support@packagecloud.io>"
 Fingerprint: XXXXXXXXXXXXX
 From       : https://packagecloud.io/netdata/netdata/gpgkey
Generating yum cache for netdata_netdata-source...

The repository is setup! You can now install packages.
[root@XXXXXXX netdata]# ls -ltr /etc/yum.repos.d/
total 36
-rw-r--r-- 1 root root  314 Nov 23 13:16 CentOS-fasttrack.repo
-rw-r--r-- 1 root root 5701 Nov 23 13:16 CentOS-Vault.repo
-rw-r--r-- 1 root root 1331 Nov 23 13:16 CentOS-Sources.repo
-rw-r--r-- 1 root root  630 Nov 23 13:16 CentOS-Media.repo
-rw-r--r-- 1 root root  649 Nov 23 13:16 CentOS-Debuginfo.repo
-rw-r--r-- 1 root root 1309 Nov 23 13:16 CentOS-CR.repo
-rw-r--r-- 1 root root 1664 Nov 23 13:16 CentOS-Base.repo
-rw-r--r-- 1 root root  547 May 13 06:18 netdata_netdata.repo
[root@XXXXXX netdata]# ls -ltr /etc/yum.repos.d/netdata_netdata.repo ^C
[root@XXXXXX netdata]# cat /etc/yum.repos.d/netdata_netdata.repo 
[netdata_netdata]
name=netdata_netdata
baseurl=https://packagecloud.io/netdata/netdata/el/7/$basearch
repo_gpgcheck=1
gpgcheck=0
enabled=1
gpgkey=https://packagecloud.io/netdata/netdata/gpgkey
sslverify=1
sslcacert=/etc/pki/tls/certs/ca-bundle.crt
metadata_expire=300

[netdata_netdata-source]
name=netdata_netdata-source
baseurl=https://packagecloud.io/netdata/netdata/el/7/SRPMS
repo_gpgcheck=1
gpgcheck=0
enabled=1
gpgkey=https://packagecloud.io/netdata/netdata/gpgkey
sslverify=1
sslcacert=/etc/pki/tls/certs/ca-bundle.crt
metadata_expire=300
[root@XXXXXXX netdata]# 
```

Basically package cloud provides scripts that the user can run and they will identify the operating system and set it up as needed.

The changes for travis to be able to build and publish to package cloud are still under `package-cloud-deployer` branch on my fork and will merge once i have finalised Travis workflows. 

This task is considered complete for now

**Comment by paulkatsoulakis** [comment]
The mechanism for containerising the build process is ready and under testing since last week.
What we did is take advantage of Travis CI and automate the process of:
1) Create an LXC container, for the distribution we wish to build RPMs on
2) Setup all required packages on that distribution
3) Run build process and generate contents for DEB package
4) Publish the resulting DEB package to packagecloud.io

Currently we intent to support packages for the following distributions:
* Ubuntu
  disco - amd64 / x86
  cosmic - amd64 / x86
  bionic - amd64 / x86
  artful - amd64 / x86

* Debian
   Buster - amd64 / x86
   Stretch - amd64 / x86
   Jessie - amd64 / x86 
   Wheezy - amd64 / x86

At the moment the CI workflow has been implemented, we are still in the process of fine tuning the build process. The idea is to reuse the currently available configurations from contrib and then there will be another iteration to get these files up to date

**Comment by gvengel** [comment]
It would be awesome if you'd support all the Ubuntu releases currently under maintenance support by Canonical, particularly the LTS releases as these see heavy adoption. Many users will stay on an older LTS release waiting until it's no longer supported before upgrading to the next LTS.

Supported LTS releases:
bionic
xenial

Supported interim releases:
disco
cosmic

For more details see the [Ubuntu Release Cycle](https://ubuntu.com/about/release-cycle).

**Comment by paulkatsoulakis** [comment]
Hello @gvengel, thanks for your info, i totally agree with you.
I totally agree with you, you should actually check out the refined list mentioned on issue #6036 

We mention:
```
Ubuntu (https://wiki.ubuntu.com/Releases)

Ubuntu 14.04.6 LTS (Trusty Tahr)
Ubuntu 16.04.6 LTS (Xenial Xerus)
Ubuntu 18.04.2 LTS (Bionic Beaver)
Ubuntu 19.04 (Disco Dingo) **latest
```

and

```
Debian (https://www.debian.org/releases/ and https://wiki.debian.org/LTS/)

Debian 7 (Wheezy)
Debian 8 (Jessie)
Debian 9 (Stretch)
Debian 10 (buster) **Not released yet
```

Let me know if that covers your comment or if i miss anything
